### PR TITLE
Simon/issue#21/validation

### PR DESCRIPTION
**Pull request by psi77**
@blaise17 How exhaustive do we want this validation to be?
For instance do we want a validator that checks integrity of obz and obf files in terms of id references?

I'm happy to add tickets for further work.

Have I got the right level of validation for simple stuff? (IsString, IsNumber, IsURL etc)

### CODE REVIEW

File: src/app/obfboard.spec.ts [hunk 1]
```diff
@@ -146,3 +145,132 @@ describe('obfboard.LoadBoardAction', () => {
...
+        rows: 2,
+        columns: 2,
+        order: [
+          [1, null],
+          [null, 2, 3]
+        ]
+      },
+      buttons: [],
+      images: [],
+      sounds: [] };
```
Comment (blaise17) [inline src/app/obfboard.spec.ts]: Not now for this review, but would it be useful to have a standard empty board json that could be used and edited per test as needed?

File: src/app/obfboard.ts [hunk 2]
```diff
@@ -79,14 +128,37 @@ export class Button {
...
+
+  @IsOptional()
+  @IsUrl()
   url: string;
+
+  @IsOptional()
+  @IsString()
   path: string;
+
   contentType: string;
```
[the remark this answers is not in the document]
Comment (psi77) [reply src/app/obfboard.ts]: I'll put a ticket in for it.....we should probably check for valid types

[the remark this answers is not in the document]
Comment (psi77) [reply src/app/obfboard.ts]: See issue#81

File: src/app/obfboard.ts [hunk 3]
```diff
@@ -162,20 +249,45 @@ export class Sound {
...
+  })
   sounds: Sound[];
+
   imageResolver: ImageResolver;
   soundResolver: SoundResolver;
 
   deserialize(input: any): OBFBoard {
     this.format = input.format;
-    this.id = String(input.id);
+    this.id = input.id && String(input.id);
```
[the remark this answers is not in the document]
Comment (psi77) [reply src/app/obfboard.ts]: Just found another one from you asking for a test for the empty image ids!

File: src/app/obfboard.spec.ts (hunk 1 above)
Comment (psi77) [reply src/app/obfboard.spec.ts]: Hmmmm dunno.  How would we amend it for a specific test?

Comment (blaise17) [reply src/app/obfboard.spec.ts]: let input = basicEmptyBoardJson();
input.grid.order = [["2", null]], [null, "1", null"]];

File: src/app/obfboard.ts [hunk 4]
```diff
@@ -1,25 +1,56 @@
...
     this.order = input.order.map(row => row.map(cell => cell || cell === 0 ? String(cell) : undefined));
-
     return this;
   }
 }
 
 export class LoadBoardAction {
+
+  @IsNotEmpty({ message: 'Load board action id must be specified'})
+  @IsString({ message: 'Load board action id must be a string'})
```
Comment (blaise17) [inline src/app/obfboard.ts]: According to the OBF spec (and examples, and our code usage) load board action id is not necessary, or relevant, or used, and is possibly misleading

Comment (blaise17) [reply src/app/obfboard.ts]: However, load board must specify one of url, dataUrl or path, so we should probably use the OneOf validator on it

File: src/app/obfboard.ts [hunk 5]
```diff
@@ -34,19 +65,37 @@ export class LoadBoardAction {
...
   borderColor: string;
   actions: string[];
   loadBoardAction: LoadBoardAction;
+
+  @IsDefined()
   parent: OBFBoard;
 
   deserialize(input: any, parent: OBFBoard): Button {
-    this.id = String(input.id);
+    this.id = input.id && String(input.id);
```
Comment (blaise17) [inline src/app/obfboard.ts]:
```suggestion
    this.id = (input.id || input.id === 0) && String(input.id);
```

Comment (blaise17) [reply src/app/obfboard.ts]: Otherwise it doesn't cast an id of 0

File: src/app/obfboard.ts [hunk 6]
```diff
@@ -79,18 +128,42 @@ export class Button {
...
+  @IsDefined()
   parent: OBFBoard;
+
+  @IsOptional()
+  @IsString()
   svgData: string;
 
   deserialize(input: any, parent: OBFBoard): Image {
-    this.id = String(input.id);
+    this.id = input.id && String(input.id);
```
Comment (blaise17) [inline src/app/obfboard.ts]: This won't work as expected if input.id is 0
```suggestion
     this.id = (input.id || input.id === 0) && String(input.id);
```

File: src/app/obfboard.ts [hunk 7]
```diff
@@ -130,16 +203,31 @@ export class Image {
...
   contentType: string;
+
   duration: number;
+
+  @IsDefined()
   parent: OBFBoard;
 
   deserialize(input: any, parent: OBFBoard): Sound {
-    this.id = String(input.id);
+    this.id = input.id && String(input.id);
```
Comment (blaise17) [inline src/app/obfboard.ts]:
```suggestion
     this.id = (input.id || input.id === 0) && String(input.id);
```
Because 0 won't cast otherwise

File: src/app/obfboard.ts [hunk 8]
```diff
@@ -162,20 +250,45 @@ export class Sound {
...
+  })
   sounds: Sound[];
+
   imageResolver: ImageResolver;
   soundResolver: SoundResolver;
 
   deserialize(input: any): OBFBoard {
     this.format = input.format;
-    this.id = String(input.id);
+    this.id = input.id && String(input.id);
```
Comment (blaise17) [inline src/app/obfboard.ts]:
```suggestion
     this.id = (input.id || input.id === 0) && String(input.id);
```
To be safe with input.id of 0

**Comment by psi77** [comment]
So what's in it to start with?

**Comment by blaise17** [comment]
Whatever is the most useful and passes minimal validation...

return
{ id: 'test', name: 'board_name', grid: { rows: 1, columns: 1, order: [
[null] ] }, buttons: [], images: [], sounds: [] };

?
It might not be useful enought to warrant it, it just takes up a lot of
space properly formatted in each test is all.

**Comment by psi77** [comment]
Should we just start putting them in files and using them as fixtures?

File: src/app/obfboard.ts [hunk 9]
```diff
@@ -34,23 +68,41 @@ export class LoadBoardAction {
...
 
   deserialize(input: any, parent: OBFBoard): Button {
-    this.id = String(input.id);
+    this.id = stringify(input.id);
     this.label = input.label;
     this.vocalization = input.vocalization;
-    this.imageId = input.image_id || input.image_id === 0 ? String(input.image_id) : undefined;
-    this.soundId = input.sound_id || input.sound_id === 0 ? String(input.sound_id) : undefined;
+    this.imageId = stringify(input.image_id);
+    this.soundId = stringify(input.sound_id);
```
Comment (blaise17) [inline src/app/obfboard.ts]: This is much nicer :)

File: src/app/custom-validation.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
+import { registerDecorator, ValidationOptions, ValidationArguments } from 'class-validator';
+
+export function OneOf(otherProperties: string[], validationOptions?: ValidationOptions) {
```
Comment (blaise17) [inline src/app/custom-validation.ts]:
```suggestion
export function OneOf(validationProperties: string[], validationOptions?: ValidationOptions) {
```

File: src/app/custom-validation.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  return function (object: Object, propertyName: string) {
+    registerDecorator({
+      name: 'OneOf',
+      target: object.constructor,
+      propertyName: propertyName,
+      constraints: [otherProperties],
```
Comment (blaise17) [inline src/app/custom-validation.ts]:
```suggestion
      constraints: [validationProperties],
```
